src/cljam/io/vcf/util.clj

```diff
@@ -176,3 +231,145 @@
           (update :info stringify-info)
           (update fmt-kw stringify-format)
           (merge (into {} (for [k sample-kws] [k (stringify-sample (v fmt-kw) (v k))])))))))
+
+(defmacro ^:private ->breakend [^String alt p1 p2 t1 t2 strand join]
+  `(let [p1# ~p1
+         p2# ~p2
+         t1# ~t1
+         t2# ~t2
+         k# (.lastIndexOf ~alt (int 58) p2#)
+         k'# (inc k#)] ;; \:
+     (when (and (< p1# k#) (< k'# p2#) (< t1# t2#))
+       (when-let [pos# (try
+                         (Long/parseLong (.substring ~alt k'# p2#))
+                         (catch NumberFormatException _# nil))]
+         {:chr (.substring ~alt p1# k#),
+          :pos pos#,
+          :bases (.substring ~alt t1# t2#),
+          :strand ~strand,
+          :join ~join}))))
+
+(defn parse-breakend
+  "Parses an ALT allele string of SVTYPE=BND.
+  Returns a map with mandatory keys:
+  - `:bases` bases that replaces the reference place
+  - `:join` `:before` or `:after`
+  and optional keys for a mate:
+  - `:chr` chromosome name of the mate sequence
+  - `:pos` genomic position of the mate sequence
+  - `:strand` strand of the mate sequence
+  Returns `nil` if input `alt` is not a breakend."
+  [^String alt]
+  (let [left-bracket (int 91) ;; \[
+        right-bracket (int 93) ;; \]
+        l1 (.indexOf alt left-bracket)
+        r1 (.indexOf alt right-bracket)]
+    (if-not (neg? l1)
+      (let [l2 (.indexOf alt left-bracket (inc l1))
+            l2' (inc l2)]
+        (when-not (or (neg? l2)
+                      (not (neg? (.indexOf alt left-bracket l2')))
+                      (not (neg? r1)))
+          (if (zero? l1)
+            (->breakend alt 1 l2 l2' (.length alt) :reverse :before)
+            (->breakend alt (inc l1) l2 0 l1 :forward :after))))
+      (if-not (neg? r1)
+        (let [r2 (.indexOf alt right-bracket (inc r1))
+              r2' (inc r2)]
+          (when-not (or (neg? r2)
+                        (not (neg? (.indexOf alt right-bracket r2'))))
+            (if (zero? r1)
+              (->breakend alt 1 r2 r2' (.length alt) :forward :before)
+              (->breakend alt (inc r1) r2 0 r1 :reverse :after))))
+        (let [dot (int 46) ;; \.
+              d (.indexOf alt dot)] ;; single breakends
+          (when-not (or (neg? d)
+                        (= 1 (.length alt))
+                        (not (neg? (.indexOf alt dot (inc d)))))
+            (if (zero? d)
+              {:bases (.substring alt 1), :join :before}
+              {:bases (.substring alt 0 d), :join :after})))))))
+
+(defn- parse-alt'
+  [^String ref ^String alt]
+  (let [r-len (.length ref)
+        a-len (.length alt)
+        eq (fn [^long r ^long a]
+             (= (Character/toUpperCase (.charAt ref r))
+                (Character/toUpperCase (.charAt alt a))))
+        [^long r ^long a] (loop [r 0 a 0]
```

Comment:
I think we don't have to define two variables (`r` and `a`) because these are bound to same value.
If you chose this implements for readability, please discard this comment.

Reply:
Thank you for pointing this out! I removed one.